File: GameLibrary/src/cc/lib/risk/RiskBoard.java
package cc.lib.risk;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import cc.lib.board.BEdge;
import cc.lib.board.BVertex;
import cc.lib.board.CustomBoard;
import cc.lib.game.Utils;

/**
 * Created by Chris Caron on 9/13/21.
 */
public class RiskBoard extends CustomBoard<BVertex, BEdge, RiskCell> {

    int [][] distMatrix = null;

    @Override
    protected RiskCell newCell(List<Integer> pts) {
        return new RiskCell(pts);
    }

    public void reset() {
        for (RiskCell cell : getCells()) {
            cell.reset();
        }
    }

    public List<Integer> getConnectedCells(RiskCell cell) {
        Set<Integer> all = new HashSet<>(cell.getConnectedCells());
        all.addAll(Utils.map(cell.getAdjCells(), idx -> idx));
        return new ArrayList<>(all);
    }

    public List<RiskCell> getAllTerritories() {
        return Utils.map(getCells(), cell -> cell);
    }

    public List<Integer> getTerritories(Army army) {
        return Utils.filter(Utils.getRangeIterator(0, getNumCells()-1), idx->getCell(idx).getOccupier()==army);
    }

    public List<Integer> getTerritories(Region region) {
        return Utils.filter(Utils.getRangeIterator(0, getNumCells()-1), idx->getCell(idx).getRegion()==region);

    }

    public void moveArmies(int fromCellIdx, int toCellIdx, int numArmies) {
        RiskCell from = getCell(fromCellIdx);
        Utils.assertTrue(from.numArmies > numArmies);
        RiskCell to = getCell(toCellIdx);
        Utils.assertTrue(to.numArmies > 0);
        from.numArmies -= numArmies;
        to.numArmies += numArmies;
    }

    private int [][] computeFloydWarshallDistanceMatrix() {
        final int numV = getNumCells();
        final int INF = Integer.MAX_VALUE/2 - 1;
        int [][] dist = new int[numV][numV];
        for (int i=0; i<numV; i++) {
            for (int ii=0; ii<numV; ii++) {
                dist[i][ii] = INF;
            }
            dist[i][i] = 0;
        }

        for (int cellIdx = 0; cellIdx<getNumCells(); cellIdx++) {
            RiskCell cell = getCell(cellIdx);
            for (int adjIdx : cell.getAllConnectedCells()) {
                dist[cellIdx][adjIdx] = dist[adjIdx][cellIdx] = 1;
            }
        }

        for (int k=0; k<numV; k++) {
            for (int i=0; i<numV; i++) {
                for (int j=0; j<numV; j++) {
                    dist[i][j] = Math.min(dist[i][k] + dist[k][j], dist[i][j]);
                }
            }
        }
        return dist;
    }


    public int getDistance(int fromCellIdx, int toCellIdx) {
        if (distMatrix == null) {
            distMatrix = computeFloydWarshallDistanceMatrix();
        }
        return distMatrix[fromCellIdx][toCellIdx];
    }
}

File: Risk/src/cc/lib/risk/RiskBoard.kt
package cc.lib.risk

import cc.lib.board.BEdge
import cc.lib.board.BVertex
import cc.lib.board.CustomBoard
import kotlin.math.min

/**
 * Created by Chris Caron on 9/13/21.
 */
class RiskBoard : CustomBoard<BVertex, BEdge, RiskCell>() {
	private val distMatrix by lazy {
		computeFloydWarshallDistanceMatrix()
	}

	override fun newCell(pts: List<Int>): RiskCell {
		return RiskCell(verts = pts)
	}

	fun reset() {
		for (cell in cells) {
			cell.reset()
		}
	}

	fun getConnectedCells(cell: RiskCell): List<Int> = HashSet<Int>().apply {
		addAll(cell.connectedCells)
		addAll(cell.adjCells)
	}.toList()

	val allTerritories by lazy {
		cells.map { cell: RiskCell -> cell }
	}

	fun getTerritories(army: Army): List<Int> {
		return Array(numCells) { it }.filter { idx -> getCell(idx).occupier == army }
	}

	fun getTerritories(region: Region): List<Int> {
		return Array(numCells) { it }.filter { idx -> getCell(idx).region == region }
	}

	fun moveArmies(fromCellIdx: Int, toCellIdx: Int, numArmies: Int) {
		val from = getCell(fromCellIdx)
		assert(from.numArmies > numArmies)
		val to = getCell(toCellIdx)
		assert(to.numArmies > 0)
		from.numArmies -= numArmies
		to.numArmies += numArmies
	}

	private fun computeFloydWarshallDistanceMatrix(): Array<IntArray> {
		val numV = numCells
		val INF = Int.MAX_VALUE / 2 - 1
		val dist = Array(numV) { IntArray(numV) }
		for (i in 0 until numV) {
			for (ii in 0 until numV) {
				dist[i][ii] = INF
			}
			dist[i][i] = 0
		}
		for (cellIdx in 0 until numCells) {
			val cell = getCell(cellIdx)
			for (adjIdx in cell.getAllConnectedCells()) {
				dist[adjIdx][cellIdx] = 1
				dist[cellIdx][adjIdx] = dist[adjIdx][cellIdx]
			}
		}
		for (k in 0 until numV) {
			for (i in 0 until numV) {
				for (j in 0 until numV) {
					dist[i][j] = min(dist[i][k] + dist[k][j], dist[i][j])
				}
			}
		}
		return dist
	}

	fun getDistance(fromCellIdx: Int, toCellIdx: Int): Int = distMatrix[fromCellIdx][toCellIdx]
}